### PR TITLE
Prompt Learning Pipeline Parallel

### PR DESCRIPTION
**Pull request by vadam5**
# What does this PR do ?

1. Adds pipeline parallel capability  to GPT p-tuning and prompt tuning
2. makes prompt learning methods compataible with GPT models trained with amp O2, 
3. switches to smaller gpt model and datasets for CI test, adding tp and pp tests

**Collection**: BigNLP

# Changelog 
- Updated MegatronGPTPromptLearning class
- Updated prompt learning config
- Moved all reduce gradient method to base megatron model
- Updated prompt learning CI tests

# Usage

```
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
- [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: examples/nlp/language_modeling/conf/megatron_gpt_prompt_learning_config.yaml [hunk 1]
```diff
@@ -11,7 +11,7 @@ trainer:
   max_epochs: 3
   max_steps: null # consumed_samples = global_step * micro_batch_size * data_parallel_size * accumulate_grad_batches
   log_every_n_steps: 10
-  val_check_interval: 200
+  val_check_interval: 1.0
```
Comment (yaoyu-33) [inline examples/nlp/language_modeling/conf/megatron_gpt_prompt_learning_config.yaml:14]: why this is changed to 1?  Run val per 1 step?

File: examples/nlp/language_modeling/conf/megatron_gpt_prompt_learning_config.yaml [hunk 2]
```diff
@@ -104,6 +105,6 @@ model:
       name: CosineAnnealing
       warmup_steps: 50
       constant_steps: 10
-      min_lr: 1e-6
+      min_lr: 0.0 # min_lr must be 0.0 if pipeline_parallel > 1
```
Comment (yaoyu-33) [inline examples/nlp/language_modeling/conf/megatron_gpt_prompt_learning_config.yaml:108]: why is this?

File: examples/nlp/language_modeling/conf/megatron_gpt_prompt_learning_config.yaml (hunk 1 above)
Comment (vadam5) [reply examples/nlp/language_modeling/conf/megatron_gpt_prompt_learning_config.yaml:14]: 1.0 runs val once per epoch. I changed this because I made the CI tests use a smaller dataset and didn't want to worry about errors saying val check interval must be less than number of steps in 1 epoch.

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_prompt_learning_model.py [hunk 3]
```diff
@@ -80,31 +84,70 @@ def __init__(self, cfg: DictConfig, trainer: Trainer):
...
+
+            if self.frozen_model.cfg.precision == 16:
+                self.float_type = torch.float16
+            elif self.frozen_model.cfg.precision == 'bf16':
+                self.float_type = torch.bfloat16
+            else:
+                raise ValueError(f'Precision {self.frozen_model.precision} Not supported with O2')
+
         # Freeze all GPT model weights for prompt-tuning/p-tuning
         self.frozen_model.freeze()
```
Comment (ericharper) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_prompt_learning_model.py:120]: We don't need to freeze the model if we're setting the LR to 0 for it's weights?

Comment (vadam5) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_prompt_learning_model.py]: Removed it

File: examples/nlp/language_modeling/conf/megatron_gpt_prompt_learning_config.yaml (hunk 2 above)
Comment (vadam5) [reply examples/nlp/language_modeling/conf/megatron_gpt_prompt_learning_config.yaml]: To keep the gpt model frozen and use DDP with pipeline parallel models, I had to make two optimizer param groups. In the frozen gpt model optimizer param groups, I force the learning rate to be 0.0 all through training so the weights that should be frozen don't get updated. If the min_lr is > 0.0, error will get raised saying min_lr > initial lr.

**Review by ericharper — DISMISSED** [review]
LGTM. Thanks!

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!
